contracts/src/L2/L2BridgeController.sol

```diff
@@ -0,0 +1,216 @@
+            _processEjection(tokenId, data);
+        }
+        
+        return this.onERC1155Received.selector;
+    }
+
+    /**
+     * @dev Default implementation of onRenew that does nothing.
+     * Can be overridden in derived contracts for custom behavior.
+     */
+    function onRenew(uint256 tokenId, uint64 expires, address /*renewedBy*/) external virtual {
+        bridge.sendMessage(BridgeEncoder.encodeRenewal(tokenId, expires));
+    }
+
+    /**
+     * @dev Default implementation of onRelinquish that does nothing.
+     * Can be overridden in derived contracts for custom behavior.
+     */
+    function onRelinquish(uint256 /* tokenId */, address /* relinquishedBy */) external virtual {
+        // Default implementation does nothing
+    }
+
+    /**
+     * @dev Validates 2LD structure and checks if label exists
+     * @param name The DNS-encoded name (must be a 2LD like "example.eth")
+     * @return tokenId The token id of the name
+     */
+    function _findAndValidateLabelStructure(
+        bytes memory name
+    ) internal view returns (uint256 tokenId) {
+        // Read the second label which should be "eth"
+        uint256 labelSize = uint8(name[0]);
+        uint256 tldOffset = 1 + labelSize;
+        uint256 tldSize = uint8(name[tldOffset]);
+        
+        // Verify TLD is "eth"
+        (bytes32 tldHash, ) = NameCoder.readLabel(name, tldOffset);
+        if (tldHash != ETH_TLD_HASH) {
+            revert InvalidTLD(name);
+        }
+        
+        // Verify the name ends after TLD (next byte should be 0)
+        if (name[tldOffset + 1 + tldSize] != 0) {
+            revert InvalidTLD(name);
+        }
+        
+        // Read the 2LD label
+        string memory label = NameUtils.readLabel(name, 0);
+        
+        // Check if the label exists in the eth registry
+        bool exists = address(registry.getSubregistry(label)) != address(0);
+        if (!exists) {
+            revert NameNotFound(name);
+        }
+
+        (tokenId, , ) = registry.getNameData(label);
+        
+        return tokenId;
+    }
+
+    /**
+     * Overrides the EjectionController._onEject function.
+     */
+    function _onEject(uint256[] memory tokenIds, TransferData[] memory transferDataArray) internal virtual override {
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            uint256 tokenId = tokenIds[i];
+            TransferData memory transferData = transferDataArray[i];
+
+            // check that the label matches the token id
+            _assertTokenIdMatchesLabel(tokenId, transferData.label);
+
+            // check that there is no more than holder of the token observer and subregistry setting roles
+            uint256 roleBitmap = ROLE_SET_TOKEN_OBSERVER | ROLE_SET_SUBREGISTRY;
+            (uint256 counts, uint256 mask) = IEnhancedAccessControl(address(registry)).getAssigneeCount(registry.getTokenIdResource(tokenId), roleBitmap);
+            if (counts & mask != roleBitmap) {
```

Comment:
A comment here on why this ensures there's exactly one holder of each of those roles would be useful. Also, we should note that we don't need to explicitly check we're the owner of those roles, since we exercise them both further down.